### PR TITLE
Support for medusa-operator

### PR DESCRIPTION
**Pull request by jeffbanks**
Modifications necessary for medusa-operator interoperability related to backup and status operations.

Effort part of [k8ssand-628](https://k8ssandra.atlassian.net/browse/K8SSAND-628).

### CODE REVIEW

**Comment by adejanovski** [comment]
@jeffbanks, I was thinking of less invasive changes, which would mostly affect the grpc server code. Would it be possible to keep the `backup_node.py` code as is and invoke it [here](https://github.com/thelastpickle/cassandra-medusa/pull/386/files#diff-0af9023e5dba3232ce866a7482e0b302d746bfa87c8040bb6c3516d481767679R50) in an async fashion, storing the future in-memory in the server? You could then have the backup status check verify the state of the future.

You'd change way less code this way, which would be safer. Let me know if I'm missing something here and my plan isn't realistic.

**Comment by jeffbanks** [comment]
> @jeffbanks, I was thinking of less invasive changes, which would mostly affect the grpc server code. 

@adejanovski I initially started down that path until I started thinking about the medusacli.py and possible future UI that may need to invoke backup outside of gRPC.  We would most likely abstract the UI with a REST call or just use the gRPC now that I think of it.  

I'm open to that if we want to go in that direction, if we don't foresee any benefit to abstracting the async to a lower level.  It will be an easy change to migrate to the gRPC server level.  We can always use gRPC for the single point of backups if the need arises.  wdyt?

**Comment by adejanovski** [comment]
> > @jeffbanks, I was thinking of less invasive changes, which would mostly affect the grpc server code.
> 
> @adejanovski I initially started down that path until I started thinking about the medusacli.py and possible future UI that may need to invoke backup outside of gRPC. We would most likely abstract the UI with a REST call or just use the gRPC now that I think of it.
> 
> I'm open to that if we want to go in that direction, if we don't foresee any benefit to abstracting the async to a lower level. It will be an easy change to migrate to the gRPC server level. We can always use gRPC for the single point of backups if the need arises. wdyt?

My opinion is that it's better not to anticipate on potential future needs and making the backup node code asynchronous does not seem to bring immediate advantages over making the call async in the grpc server. I think gRPC is our remote invocation tool of choice for the foreseeable future anyway.
I'm verry sorry you did all this work and I'm now asking you to do something else 😕  If you could just implement the thread pool in the grpc server, it would be awesome.

Thanks!

**Comment by jeffbanks** [comment]
> I think gRPC is our remote invocation tool of choice for the foreseeable future anyway.
Good to know, that makes sense to me.

> I'm verry sorry you did all this work
No worries, not really that much new logic was added, only the backup tracking and thread demarcation will be migrated over to the `gRPC server`.  The majority of code changes in `backup_node` was for testability and thread start with tracking, and was essentially the same backup logic we use today.

**Comment by jeffbanks** [comment]
**Updates:**
- Async backup for gRPC calls only; modification requested in initial PR review.
- Unit and integration test updates to support status checks for async backup.
- Added `backup_node_test` for unit-level coverage for async capabilities.
- Reduced overall unit-test runtime from 1+ min. to ~10 sec. -- root cause: DNS resolver usage in unit-tests.

**Review by adejanovski — COMMENTED** [review]
I still need to do some testing, but the PR looks great so far!

Just a small concern on moving `BackupMan` to its own source file and avoiding static references, and a few loggings that should be removed or changed of level.

Thanks!

**Comment by jeffbanks** [comment]
@adejanovski I am getting back to this effort after switching priorities on other projects!  Sorry about the delay on this one.

File: medusa/storage/__init__.py [hunk 1]
```diff
@@ -403,9 +403,9 @@ def latest_complete_cluster_backup(self, backup_index=None):
         return last_finished
 
     def get_cluster_backup(self, backup_name):
-        for cluster_backup in self.list_cluster_backups():
-            if cluster_backup.name == backup_name:
-                return cluster_backup
+        for cb in self.list_cluster_backups():
```
[the remark this answers is not in the document]
Comment (jeffbanks) [reply medusa/storage/__init__.py]: Shadow warning I was fixing here.  Changing it back to the full name.

File: medusa/backup_node.py [hunk 2]
```diff
@@ -156,69 +228,53 @@ def stagger(fqdn, storage, tokenmap):
...
-def main(config, backup_name_arg, stagger_time, enable_md5_checks_flag, mode):
+# Kicks off the node backup unit of work and registers for backup queries.
+# No return value, throws back exception for failed kickoff.
+def handle_backup(config, backup_name_arg, stagger_time, enable_md5_checks_flag, mode):
     start = datetime.datetime.now()
     backup_name = backup_name_arg or start.strftime('%Y%m%d%H%M')
     monitoring = Monitoring(config=config.monitoring)
 
     try:
+        logging.info("Inside handle_backup")
```
[the remark this answers is not in the document]
Comment (jeffbanks) [reply medusa/backup_node.py:248]: Changed to debug level and added the mode that is being used.  I think this is a good milestone in the log for tracking down any issues.

File: medusa/backup_node.py [hunk 3]
```diff
@@ -156,69 +228,53 @@ def stagger(fqdn, storage, tokenmap):
...
         storage = Storage(config=config.storage)
         cassandra = Cassandra(config)
 
         differential_mode = False
         if mode == "differential":
             differential_mode = True
 
+        logging.debug("Calling get_node_backup")
```
[the remark this answers is not in the document]
Comment (jeffbanks) [reply medusa/backup_node.py:256]: This one seems redundant, I'll remove it.

File: medusa/backup_node.py [hunk 4]
```diff
@@ -156,69 +228,53 @@ def stagger(fqdn, storage, tokenmap):
...
-
-        actual_start = datetime.datetime.now()
-
-        enable_md5 = enable_md5_checks_flag or medusa.utils.evaluate_boolean(config.checks.enable_md5_checks)
-        num_files, node_backup_cache = do_backup(
-            cassandra, node_backup, storage, differential_mode, enable_md5, config, backup_name)
-
-        end = datetime.datetime.now()
-        actual_backup_duration = end - actual_start
+        logging.info("Calling start_backup")
```
[the remark this answers is not in the document]
Comment (jeffbanks) [reply medusa/backup_node.py:265]: Adjusting this milestone log to something like: `starting backup using stagger: {} mode: {} name: {}` as this is invoked after the backup preparations are in-place.

File: medusa/restore_cluster.py [hunk 5]
```diff
@@ -261,15 +262,17 @@ def _get_seeds_fqdn(self):
...
         return seeds
 
     def _populate_hostmap(self):
         with open(self.host_list, 'r') as f:
             for line in f.readlines():
                 seed, target, source = line.replace('\n', '').split(self.config.storage.host_file_separator)
                 # in python, bool('False') evaluates to True. Need to test the membership as below
-                self.host_map[self.fqdn_resolver.resolve_fqdn(target.strip())] \
-                    = {'source': [self.fqdn_resolver.resolve_fqdn(source.strip())], 'seed': seed in ['True']}
+                target_resolved = self.fqdn_resolver.resolve_fqdn(target.strip())
```
[the remark this answers is not in the document]
Comment (jeffbanks) [reply medusa/restore_cluster.py:297]: That is correct, the second call `resolve` benefits from the first lookup in that case.

File: medusa/backup_node.py [hunk 6]
```diff
@@ -118,9 +118,81 @@ def _make_manifest_object(self, path_prefix, cached_item):
         return ManifestObject('{}{}'.format(path_prefix, cached_item['path']), cached_item['size'], cached_item['MD5'])
 
 
+class BackupMan:
```
[the remark this answers is not in the document]
Comment (jeffbanks) [reply medusa/backup_node.py]: Yes, makes sense for other references as well outside of these two files.  I've moved the `backup manager` into its own file.

**Comment by jeffbanks** [comment]
@adejanovski let me know if you think we should further reduce the logging.  I found most of the changes helpful at `info` but have moved most to `debug` to reduce verbosity in the logs.

File: requirements.txt [hunk 7]
```diff
@@ -15,6 +15,66 @@ ssh-python>=0.6.0
...
+Pygments~=2.2.0
+Pillow~=5.1.0
+roman~=2.0.0
+simplejson~=3.13.2
+Twisted~=17.9.0
+virtualenv~=20.4.3
+distlib~=0.3.1
+appdirs~=1.4.4
+fido2~=0.9.1
+zipp~=3.4.1
```
Comment (adejanovski) [inline requirements.txt:80]: Why did you add all these as requirements? It looks like some of them broke the debian package builds, and some of them should definitely be kept out of here such as `google` or `boto3` for example, since they're backend specific.
We already have `requirements-s3.txt`  or `requirements-gcs.txt` which contain these dependencies through higher level ones.

Comment (jeffbanks) [reply requirements.txt:80]: Good catch.   Picked up those during some local testing.  I need to use the pip `freeze` option before local changes.

**Review by adejanovski — CHANGES_REQUESTED** [review]
I've initiated [a medusa-operator branch](https://github.com/k8ssandra/medusa-operator/compare/master...adejanovski:async-backup-completion-checks?expand=1) to test the changes you made here.

So far it showed two things:

- There's a problem when `BackupStatus` is called multiple times for the same pod. The first call works as expected (showing a NotFound error as the backup doesn't exist yet):
```
2021-11-30T12:57:10.722Z	INFO	controllers.CassandraBackup	Checking backup status	{"CassandraPod": "k8ssandra-dc1-default-sts-2"}
2021-11-30T12:57:10.764Z	ERROR	controllers.CassandraBackup	failed hasAsyncBackupStatus	{"Pod": "k8ssandra-dc1-default-sts-2", "error": "rpc error: code = NotFound desc = backup <backup5> does not exist"}
```

But subsequent calls show the following error:

```
2021-11-30T12:57:13.389Z	ERROR	controllers.CassandraBackup	failed hasAsyncBackupStatus	{"Pod": "k8ssandra-dc1-default-sts-1", "error": "rpc error: code = Unknown desc = Exception calling application: Assignment not allowed to repeated field \"finishedNodes\" in protocol message object."}
github.com/k8ssandra/medusa-operator/controllers.(*CassandraBackupReconciler).Reconcile.func1.1
	/workspace/controllers/cassandrabackup_controller.go:207
2021-11-30T12:57:13.389Z	INFO	controllers.CassandraBackup	Async backups are NOT available
```

I'm not sure what's causing this.

- We would need to keep the `Backup()` operation as blocking to maintain backwards compatibility with prior versions of medusa-operator, and create an `AsyncBackup()` operation which will return immediately and require `BackupStatus()` calls to check for progress. This way we can release a new version of Medusa without requiring an upgrade of medusa-operator (and we possibly have people using the GRPC server even if they don't use medusa-operator).

**Comment by jeffbanks** [comment]
@adejanovski 

I'm updating this PR to support both `async` and `sync` backups for use by various gRPC clients. 
 
Regarding the error you are seeing, I'm also adding additional gRPC integration tests to drill-down on the backup results a bit.  
This should help us shed some light on the problem and assist with future `medusa-operator` integration.

**Comment by jeffbanks** [comment]
Latest changes:

* K8SSAND-1074 Resolves assignment to repeated field error.
* K8SSAND-1075 Includes retention of sync backup ops.
* Integration tests added to support async and non-async backup paths.
* Generalized StatusType enum for use by various gRPC responses.
* Added UNKNOWN StatusType for default state of registered backups (async and sync).
* Server and client exception handling updates.
* Detailed error conditions added for gRPC client.
* Refactored complexity of specific methods and decoupling of logic in gRPC server.
* BackupMan updated to clean and restart backup of same name; self-heal instead of discontinue.
* BackupMan updated to handle tracking of backup status for all types (not just async).
* Overall status included as part of GetBackups gRPC method.

File: medusa/medusacli.py [hunk 8]
```diff
@@ -17,6 +17,8 @@
 from gevent import monkey
 
 import medusa.utils
+from medusa import backup_node
+from medusa.backup_manager import BackupMan
```
Comment (adejanovski) [inline medusa/medusacli.py:21]: These two line should be placed after `monkey.patch_all()` otherwise it'll break some stuff in the orchestration of backups.
Also `from medusa import backup_node` should replace `import medusa.backup_node` on line 40 which isn't necessary anymore.

**Review by adejanovski — CHANGES_REQUESTED** [review]
There's one change to make in the order of imports so that monkey patching still works, then we'll be good to merge.

**Review by adejanovski — APPROVED** [review]
Looks good now 👍 
Feel free to squash and merge the PR now @jeffbanks
